docs/proposals/image-signing.md

```diff
@@ -0,0 +1,129 @@
+define policy, e.g. trusted keys, applicable when deciding whether to accept an image, or
+individual signatures of that image, as valid:
+
+```json
+{
+  "default": [{"type": "reject"}],
+  "transports": {
+    "atomic": {
+      "172.30.62.231:5000": [
+      {
+        "type": "signedBy",
+        "keyType": "GPGKeys",
+        "keyPath": "/keys/public.gpg"
+      }
+      ]
+    },
+    "docker": {
+      "docker.io/openshift": [{"type": "insecureAcceptAnything"}]
+    }
+  }
+}
+```
+
+In this example, the default policy is set to reject all Docker images by default and
+whitelist only images that are pushed in the OpenShift Integrated Registry and signed
+using the public GPG key.
+It also whitelists all Docker images from the OpenShift DockerHub repository.
+
+To sign an Docker image, you have to execute the following command:
+
+```console
+$ skopeo --tls-verify=false copy --sign-by user@example.com docker://docker.io/myrepo/image:latest \
+  atomic:172.30.74.246:5000/mynamespce/image:latest
+```
+
+When this command is executed, skopeo will ask you to enter the password the
+*user@example.com* GPG key, sign the local image and push the signed image into OpenShift
+Integrated Registry.
+
+### Kubernetes `ImagePolicyWebhook` Admission Controller
+
+Kubernetes provides the
+[ImagePolicyWebhook](http://kubernetes.io/docs/admin/admission-controllers/#imagepolicywebhook)
+to allow a backend webhook to make admission decisions about whether an image specified by
+the container inside a Pod is allowed to run in the cluster or not.
+
+To configure this you have to specify a backed HTTPS service that Kubernetes will send the
+`ImageReview` objects as a payload, expecting the `ImageReview` in return with the
+*status* field set to indicate allowance or rejection of the image(s).
+
+## Assumptions and Constraints
+
+The assumption is that we provide a HTTPS backend that implements the `ImagePolicyWebhook`
+by making calls into Skopeo [go library](https://github.com/containers/image) to evaluate
+the policy.json file provided by the cluster admin.
+
+Kubernetes requirements for the ImagePolicyWebhook backend are:
+
+* The admission controller config file (`--admission-controller-config-file`) to set configuration options for
+  the behavior of the backend
+* The admission config file must reference a kubeconfig formatted file which sets up the connection
+  to the backend
+* The backend must communicate over TLS
+
+The image verification backend requirements are:
+
+* Policy file must be provided
+* All public GPG keys must be available for the the backend server
+* The backend must have view access to all images in the registry specified in policy file
+
+The assumption is then that we will run the backend HTTPS server as a part of OpenShift
+infrastructure, similar to how we run Docker Registry or the HAProxy router.
+
+### Policy
+
+The policy must be provided in order for Skopeo to match the provided image with the rules
```

Comment:
I'd like to see what this would look like. We need this to be really flexible. I would be concerned embedding into master-config the full signature policy for the entire cluster to be unwieldy.